### DISCUSSION TITLE
Analyzing and finetuning a Lora

### DISCUSSION THREAD

**Discussion by derspanier1** [vladmandic/sdnext#1093] [General]
**Edit: I suggest to do this on the base 1.5 model if you want it tuned with the highest combability overall, and actually make 2 or 3 fine detail plots to really find the spots that hit the nail. i always focus on the eyes.** 

Due to demand and because sharing is caring, here comes my workflow of how I get the most out of a Lora, make it act more predictable and give more control. I am not claiming to have the ultimate wisdom, just a little bit of insight. Also: English 2nd no wait, 3rd language so it might sound krautisch here and there ;D

okay what is needed:

Lora Block Weight Extension 
https://github.com/hako-mikan/sd-webui-lora-block-weight

X/Y/Z Plot.

Let's begin. After installing the Extension and opening it we are greeted with the following:
![New Project](https://github.com/vladmandic/automatic/assets/117949621/b1a0bdd5-55ca-4d6f-a971-cd5cb80cbe18)

The upper field says:
NONE,ALL,INS,IND,INALL,MIDD,OUTD,OUTS,OUTALL,ALL0.5

These are containers that are defined in the lower field:

NONE:0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
ALL:1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1
INS:1,1,1,1,0,0,0,0,0,0,0,0,0,0,0,0,0
IND:1,0,0,0,1,1,1,0,0,0,0,0,0,0,0,0,0
INALL:1,1,1,1,1,1,1,0,0,0,0,0,0,0,0,0,0
MIDD:1,0,0,0,1,1,1,1,1,1,1,1,0,0,0,0,0
OUTD:1,0,0,0,0,0,0,0,1,1,1,1,0,0,0,0,0
OUTS:1,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1
OUTALL:1,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1
ALL0.5:0.5,0.5,0.5,0.5,0.5,0.5,0.5,0.5,0.5,0.5,0.5,0.5,0.5,0.5,0.5,0.5,0.5

These are basically presets that one can use to test the blocks inside of a lora. I have a few more presets in the screenshot that I added after lots of testing and some GPT help.
My Edit Presets:

NONE,ALL,INS,IND,INALL,MIDD,OUTD,OUTS,OUTALL,ALL0.5,TT,SET1,SET2,SET3,SET4

NONE:0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
ALL:1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1
INS:1,1,1,1,0,0,0,0,0,0,0,0,0,0,0,0,0
IND:1,0,0,0,1,1,1,0,0,0,0,0,0,0,0,0,0
INALL:1,1,1,1,1,1,1,0,0,0,0,0,0,0,0,0,0
MIDD:1,0,0,0,1,1,1,1,1,1,1,1,0,0,0,0,0
OUTD:1,0,0,0,0,0,0,0,1,1,1,1,0,0,0,0,0
OUTS:1,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1
OUTALL:1,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1
ALL0.5:0.5,0.5,0.5,0.5,0.5,0.5,0.5,0.5,0.5,0.5,0.5,0.5,0.5,0.5,0.5,0.5,0.5
TT:1,0.2,0.2,0.2,0.5,0.5,0.5,0.5,1,1,1,1,0.5,0.5,0.5,0.5,0.5
SET1:0.2,0.4,0.6,0.8,0.5,0.5,0.5,0.5,0.2,0.2,0.2,0.2,0.8,0.8,0.8,0.8,0.8
SET2:0.9,0.8,0.7,0.6,0.5,0.4,0.3,0.2,0.1,0.2,0.3,0.4,0.5,0.6,0.7,0.8,0.9
SET3:0.1,0.1,0.1,0.1,0.2,0.2,0.2,0.2,0.9,0.9,0.9,0.9,0.8,0.8,0.8,0.8,0.8
SET4:0.1,0.1,0.1,0.1,0.2,0.2,0.2,0.2,0.9,0.9,0.9,0.9,0.8,0.8,0.8,0.8,0.7

Now to the fun part.
Chose the Lora of your likening and whatever checkpoint you want to tune it to.
![New Project (1)](https://github.com/vladmandic/automatic/assets/117949621/5a6b98f1-8146-4dc7-b10c-3b7ed12b2ce1)

In this example I am tuning the Dafoe Lora, and using Dreamlike Diffusion because my "pick something for me" wildcard told me so lol. My First step always is to check the metadata of the lora via extra networks, copy the metadata, throw it in GPT to analyze for trigger frequency, and then build myself a structure that has randomness but is also strict on keywords. Also note the resolution i chose (384x512) because it has the highest value in the metadata.
That's how i ended with:

`{Wearing a black shirt and a suit|In a black suit and tie|Sitting on a chair in a black shirt|Posing for the camera in a suit|Wearing a black leather jacket|In a black shirt and tie, sitting at a table|Sitting in a black jacket|With a black shirt and a leather jacket|In a black shirt and tie, posing for a picture},{With a smiling face|With a big smile on his face|Smiling at the camera|Sitting with a smile on his face|Smiling for the camera}`

1 item clothing|and 1 item smiling

Prompt:

`dreamlikeart(best quality:{1.0|1.1|1.2|1.3|1.4}),(masterpiece:{1.0|1.1|1.2|1.3|1.4}),8K,portrait of <lora:Willem_Dafoev1:1.0> willedaf, {Wearing a black shirt and a suit|In a black suit and tie|Sitting on a chair in a black shirt|Posing for the camera in a suit|Wearing a black leather jacket|In a black shirt and tie, sitting at a table|Sittingt in a black jacket|With a black shirt and a leather jacket|In a black shirt and tie, posing for a picture},{With a smiling face|With a big smile on his face|Smiling at the camera|Sitting with a smile on his face|Smiling for the camera},an oil painting by {3$$__mine/Art_Artists1__} and Studio Ghibli,in {(__mine/Mod_VideoGame__:0.5)|(__mine/Mod_Movie__:0.5)},{1$$__mine/Mod_Trending__},Intricate,High Detail,Sharp focus,photo-realistic`

model trigger word, then some best quality masterpiece quackery, Lora plus the added triggers, then i pull 3 random artists out of a wildcard, then choose between either 1 item of the videogame artstyle wildcard or 1 item of the movie artstyle wildcard, 1 item of the trending on blabla wildcard, and some other detail stuff. once all of that gets resolved i end up under 75 tokens.

Negative:

`({EasyNegative|Asian-Less-Neg|bad-artist|bad-artist-anime|bad-hands-5|bad-image-v2-39000|bad-picture-chill-75v|bad_pictures|bad_prompt|bad_prompt_version2|bad_quality|FloxiQualityneg|negative_hand-neg|ng_deepnegative_v1_75t|photozoov15|qualityzoov15|renderzoov15|structurezoov15|Style-Autumn-neg|style-empire-neg|Style-Glorious-neg|style-nebmagic-neg|style-psycho-neg|Style-Winter-neg|verybadimagenegative_v1|vile_prompt3}:{0.8|0.9|1.0})`

randomly choose 1 negative embedding and then either 0.8,0.9 or 1.0 strength

Tuning The Lora:

`<lora:Willem_Dafoev1:1.0>`
This is the standard format/syntax. What we do now is take this and prepare it for the plotter:
`<lora:Willem_Dafoev1:0.25:ALL>`
I choose 0.25 (strength) because normally there is almost no influence of the Lora on the end result there. but if you wanna safe time go with 0.5. The "ALL" calls on the presets from the Block Weights.

X/Y/Z Plot:
Open the script, and choose for X and Y:
**Prompt S/R** 

X: ALL,INS,IND,INALL,MIDD,OUTD,OUTS,OUTALL,ALL0.5 or ALL,INS,IND,INALL,MIDD,OUTD,OUTS,OUTALL,ALL0.5,TT,SET1,SET2,SET3,SET4 with my additional presets

Y: 0.25,0.5,0.75,1.0 
you could add much more in-between values, important is that the first number is identical to the number in the lora (Strength).

click generate

it will now start running and test the blocks inside the Lora according to the presets, while incrementally raising the strength on the y axis. Typically things start to get close around the 0.75 mark, but every Lora is different. Once the plot is done you can exactly identify where the Lora has an impact. here is the finished plot:

https://drive.google.com/file/d/1l8zfZhTDjUMeNFX26Z41GLYinuKNVENB/view?usp=share_link

as you can see i marked the areas that to me have the highest convergence. so the values would be:

`{1.0:ALL|1.07:MIDD|1.0:OUTD|1.0:ALL0.5|1.15:TT}`

now we have locked down the areas where the action happens. what do then is take this and put it like this:

`<lora:Willem_Dafoev1:{1.0:ALL|1.07:MIDD|1.0:OUTD|1.0:ALL0.5|1.15:TT}>`

now every generation one of the "good values" gets randomly selected, making sure there remains variation, so the final prompt would be:

`dreamlikeart(best quality:{1.0|1.1|1.2|1.3|1.4}),(masterpiece:{1.0|1.1|1.2|1.3|1.4}),8K,portrait of <lora:Willem_Dafoev1:{1.0:ALL|1.07:MIDD|1.0:OUTD|1.0:ALL0.5|1.15:TT}> willedaf, {Wearing a black shirt and a suit|In a black suit and tie|Sitting on a chair in a black shirt|Posing for the camera in a suit|Wearing a black leather jacket|In a black shirt and tie, sitting at a table|Sittingt in a black jacket|With a black shirt and a leather jacket|In a black shirt and tie, posing for a picture},{With a smiling face|With a big smile on his face|Smiling at the camera|Sitting with a smile on his face|Smiling for the camera},an oil painting by {3$$__mine/Art_Artists1__} and Studio Ghibli,in {(__mine/Mod_VideoGame__:0.5)|(__mine/Mod_Movie__:0.5)},{1$$__mine/Mod_Trending__},Intricate,High Detail,Sharp focus,photo-realistic`

turn of plot, highres fix on:
`Steps: 35, Sampler: DPM++ 2M Karras, CFG scale: 6, Seed: 1162229108, Size: 384x512`

first result:
![grid-4114-1162229108-dreamlikeart(best quality_1 1),(masterpiece_1 4),8K,portrait of _lora_Willem_Dafoev1_1 0_ALL_ willedaf, In a black shirt and tie](https://github.com/vladmandic/automatic/assets/117949621/f1481b24-634c-4ae4-9ba3-4f75d32fff18)

thats it. if there are questions, feel free.

**Comment by Tillerz** [comment]
That's awesome!

**Reply by drphero** [reply]
Would you be willing to share your prompts for ChatGPT?

**Reply by derspanier1** [reply]
@drphero well in this case i just ask for specific stuff like "analyze this that, convert into this syntax" and so on, but i have a prompt for GPT that produces nice stable diffusion prompts:

```
You are a master artist, well-versed and artistic terminology with a vast vocabulary for being able to describe visually things that you see.  I'm going to give you a series of prompts that I use in a program called stable diffusion to do image generation.   I would like you to analyze the styles of these prompts, the sentence structures, how they're laid out and the common pattern between all of them. The sample prompts may have special formatting in the form of parentheses with a ":" and a number between 0 and 2 with a decimal point, for example (symmetrical Easter theme:1.3) or (cinematic:1.4) or (symmetrical:1.2).  This format should be used to emphasize or subdue parts of the description that would be important to the presentation of the image. Above 1 emphasizes whatever is in the parentheses, and anything below 1, de-emphasizes whatever is in the parentheses. do not use the word "emphasis" when utilizing this format. 

You will be asked to generate prompts based on your analysis. I want you to take note that the image generation software pays more attention to what's at the beginning of the prompt and that attention declines the closest to the end of the positive Prompt that you get.  The formatting may look something like this: [main character or focus of the image], [lighting style, shot style, and color], [other aesthetics like mood, emotion, peripheral subject matter], [possible artists or art styles]

Prompting:
Order matters - words near the front of your prompt are weighted more heavily than the things at the end of your prompt.

Most prompts should have the following at the beginning: "((best quality)), ((masterpiece)), (detailed), " 

When using an artist name or artist reference always use the format " in the style of [ artist name]", for example " in the style of Greg Rutkowski" or  " in the styles of Michelangelo and Leonardo da Vinci"

[Example low, medium and high quality prompts]:
[low detail prompt]:portrait of a futuristic beautiful woman and a futuristic city, deep vivid colors

[medium detail prompt]:modern alchemists lab drawn with watercolor pencil, magic potions, remedies, vials and concoctions, vivid colors, bubbling, bursting, full of life, organic, oily, bones, dark, scary

[High detailed, high quality prompt]: (best quality:1.4), (masterpiece:1.4), (detailed:1.3), 8K, portrait, elven queen in a lush forest, shimmering golden gown, jeweled tiara, enigmatic smile, surrounded by ethereal glowing fauna, (magic-infused:1.4), vivid colors, intricate foliage patterns, chiaroscuro lighting, (Pre-Raphaelite art style:1.2)

[Sample positive prompts to analyze]:
(best quality:1.4), (masterpiece:1.4), (detailed:1.3), 8K, underwater cityscape, vibrant coral reefs, diverse marine life, glowing bioluminescent organisms, (ancient ruins:1.3), mysterious, dreamlike atmosphere, crystal clear water, sun rays penetrating, (impressionist style:1.2), Monet and Van Gogh influences, wide-angle shot

(best quality:1.4), (masterpiece:1.4), (detailed:1.3), 8K, Victorian steampunk city, bustling streets, eclectic mix of fashion, steam-powered vehicles, airships overhead, industrial architecture, smoky atmosphere, (sepia-toned:1.4), detailed textures, H.R. Giger and Jules Verne inspirations, aerial view

(best quality:1.4), (masterpiece:1.4), (detailed:1.3), 8K, enchanted forest, fantastical creatures, mystical beings, vivid flora, towering trees, (sparkling waterfalls:1.3), vibrant colors, ethereal atmosphere, (magic-infused:1.4), Frazetta and Yoshitaka Amano styles, low-angle perspective

(best quality:1.4), (masterpiece:1.4), (detailed:1.3), 8K, cyberpunk metropolis, neon-lit streets, diverse characters, futuristic vehicles, towering skyscrapers, holographic advertisements, Blade Runner and Akira inspirations, (rain-soaked streets:1.4), wide-angle panoramic shot

(best quality:1.4), (masterpiece:1.4), (detailed:1.3), 8K, surreal landscape, floating islands, twisted dimensions, bizarre flora and fauna, (gravity-defying:1.3), Dali and Magritte influences, vibrant colors, otherworldly lighting, fisheye lens perspective

(best quality:1.4), (masterpiece:1.4), (detailed:1.3), 8K, dystopian desert wasteland, post-apocalyptic survivors, makeshift vehicles, scavenged attire, harsh climate, (sandstorms:1.3), desolate atmosphere, Mad Max and Fallout influences, high-angle shot

(best quality:1.4), (masterpiece:1.4), (detailed:1.3), 8K, whimsical candyland, sugary landscapes, confectionary creatures, (gumdrop forests:1.3), chocolate rivers, vibrant colors, playful atmosphere, (nostalgic childhood memories:1.4), Dr. Seuss and Tim Burton inspirations, aerial view

(best quality:1.4), (masterpiece:1.4), (detailed:1.3), 8K, medieval castle, knightly tournaments, bustling marketplace, intricate tapestries, (stained glass windows:1.3), vivid colors, historical accuracy, romantic atmosphere, Ivan Bilibin and Arthur Rackham influences, bird's-eye view

(best quality:1.4), (masterpiece:1.4), (detailed:1.3), 8K, futuristic space station, advanced technology, diverse alien species, interstellar ships, cosmic vistas, Star Wars and Star Trek inspirations, (epic battles:1.4), vivid colors, dynamic lighting, wide-angle shot

You will be asked to generate prompts based on the instructions above.  You will provide a prompt based on the provided information every time a user instructs you, and in those prompts I would also like you to include a description of the camera angle used to get the shot.  
You do not need to report on your analysis, please just respond with, "Tell me what you want to see.".

Please write in English language.
```

**Reply by derspanier1** [reply]
and if you mean the prompt in the example tuning, thats all learned by testing and spending a boatload of time writing wildcards. but even that can be done now in seconds by GPT. Civitai has some good wildcards in the "wildcard" section.

**Comment by drphero** [comment]
Thanks a lot for the info!

**Reply by derspanier1** [reply]
a pleasure.

**Comment by vladmandic** [comment]
@derspanier1 btw, since your doing your magic analyzing loras, any such hints for training them? :)

**Reply by derspanier1** [reply]
depends on what you want to train ?
single subject ? style ? 
well, give me a day and ill write something up regarding what works for me.

**Reply by vladmandic** [reply]
thanks - and absolutely no rush.
i was more curious about advanced network parameters when training - if you use any?
nowadays, there are sooo many options/params and most of training how-tos are really basic and go into image preprocessing/etc, but nobody really talks about network params.

**Comment by derspanier1** [comment]
[wildcards.zip](https://github.com/vladmandic/automatic/files/11525675/wildcards.zip)
until then: my wildcards. there is a lot of blood sweat and tears in there. treat them with kindness :D

**Reply by Tillerz** [reply]
`__clothing__[::1]` what's that about the `[::1]`, is that default stable diffusion or is that part of the replacement of the extension that replaces `__clothing__`? :)

**Reply by derspanier1** [reply]
@Tillerz i´ll have a look at it later, this could be something that i overlooked to remove while experimenting, but i have to check. i am in the process of calibrating hundreds of loras and then build a wildcard system that pulls in the calibrated loras and this weird looking syntax there might be some "5 am garble remnant"

**Comment by derspanier1** [comment]
Sorry i havent been active the last days as i am being inundated with work rn, but i´ll get into LoRA training as soon as this client is done.

in the meantime something for someone to think about that isnt as stupid as i am with maths and logic n sh1t.
I thought to myself: why not calculate this stuff with GPT ?

This is what i have rn:
```
Low-Rank Adaptation of Large Language Models (LoRA) is a training method that accelerates the training of large models while consuming less memory. It adds pairs of rank-decomposition weight matrices to existing weights, and only trains those newly added weights. This has a couple of advantages: LoRA is not only limited to attention layers. This is why it’s common to just add the LoRA weights to the attention layers of a model.

Base Information:
Syntax: <lora:loraname:weight:blockweights> or <lora:loraname:weight:preset>

The format is as follows:
Presets:ALL,MIDD,OUTALL,ALL0.5,TT,SET1,SET4
Preset:Blockweights:
ALL:1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0
MIDD:1.0,0,0,0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,0,0,0,0,0
OUTALL:1.0,0,0,0,0,0,0,0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0
ALL0.5:0.5,0.5,0.5,0.5,0.5,0.5,0.5,0.5,0.5,0.5,0.5,0.5,0.5,0.5,0.5,0.5,0.5
TT:1.0,0.2,0.2,0.2,0.5,0.5,0.5,0.5,1.0,1.0,1.0,1.0,0.5,0.5,0.5,0.5,0.5
SET1:0.2,0.4,0.6,0.8,0.5,0.5,0.5,0.5,0.2,0.2,0.2,0.2,0.8,0.8,0.8,0.8,0.8
SET4:0.1,0.1,0.1,0.1,0.2,0.2,0.2,0.2,0.9,0.9,0.9,0.9,0.8,0.8,0.8,0.8,0.7
The 17 Values in each preset are the corresponding blockweights, separated by a comma. The single block range is from 0 to 1. The order is important, intial blockweight values influece differently than middle ones or final ones.
Example: Preset MIDD:1,0,0,0,1,1,1,1,1,1,1,1,0,0,0,0,0 has the value 1 at the beginning, telling that the first block has full influence. then it is followed by a few 0 values, telling you no influence is on those blocks. Then it is followed by a few 1 values, telling that the mid range blocks have full influence. Then it closes with a few more 0 values, telling you the end blocks have no influence.

Additional Parameter:
Weight: Range from 0 to 2 with one decimal point, for example: 1.1 or 0.35.

I will provide you with a Lora tag with a specific weight and a preset, for example: <lora:Greenjacket:0.75:TT>.

Instructions Start:

1: Parse the Input: First, parse the input string using the format: <lora:loraname:weight:preset>. This should yield four values: the prefix ('lora'), the name of the Lora, the weight, and the preset used.

2: Interpret the Preset: Refer to the preset list provided in the base information to understand the selected preset. This preset corresponds to an array of blockweights, which signifies the level of influence applied to each block in the model.

3: Calculate Influence: Calculate the desired influence for each block by multiplying the blockweight by the provided weight.

4: For each preset, for each block, calculate the required weight that would yield the desired influence. This will give you a set of weights for each preset.

5: Instead of averaging these weights, calculate the geometric mean. This requires multiplying all the weights together and then taking the nth root (where n is the number of blocks).

6: If this weight falls within the valid range (0 to 2), this preset and weight combination is a valid alternative.

7: Repeat this for all presets to generate a list of alternative configurations and present them in the format {weight:preset|weight:preset}.

Instructions End

<lora:depths-fC_v1.1:0.85:TT>
```

of course it does not work, maybe someone of the truely smart ones around here can come up with something that could approximate closely.

i am also working on a gpt prompt optimizer that is already working really nice, and i am working on a gpt prompt gen. ill share all of this when its good enough. Good day to everyone.
